### PR TITLE
New blocks: measurements and actuation motor-side and RelativeTransform

### PR DESCRIPTION
**Pull request by diegoferigo**
This PR is in preparation of release `4` of WB-Toolbox, and covers most of the remaining tasks of [Milestone#3](https://github.com/robotology/wb-toolbox/milestone/3).

The main new features are:
- New *Get Motor Measurement* block
- New *Set Motor References* block
- New *Relative Transform* block
- Bugfixing related to these classes

Particularly, this PR:

- Fixes https://github.com/robotology/wb-toolbox/issues/71 **Relative transformation block**
- Fixes https://github.com/robotology/wb-toolbox/issues/75 **Unit of 'Reference speed' measurement in 'Set references' block**
- Fixes https://github.com/robotology/wb-toolbox/issues/80 **GetLimit block shows wrong max velocity value**
- Fixes https://github.com/robotology/wb-toolbox/issues/93 **GetControlReferences block**
- Fixes https://github.com/robotology/wb-toolbox/issues/102 **Problem when using VOCAB_CM_POSITION control mode**
- Fixes https://github.com/robotology/wb-toolbox/issues/104 **Motor positions required in the JointsPosition block**
- Fixes https://github.com/robotology/wb-toolbox/issues/105 **Read motor PWM**
- Fixes https://github.com/robotology/wb-toolbox/issues/106 **Read motor currents**

cc @traversaro @S-Dafarra @vibhoraggarwal @nunoguedelha @fiorisi
Can you please try this branch and give me feedback?

### CODE REVIEW

**Comment by S-Dafarra** [comment]
cc @gabrielenava

File: toolbox/src/RelativeTransform.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+    // Create a Signal with an identity base
+    auto fakeBase = Signal(Signal::DataFormat::CONTIGUOUS,
+                           DataType::DOUBLE,
+                           /*isConst=*/false);
+    fakeBase.setWidth(16);
+    fakeBase.initializeBufferFromContiguous(
+        iDynTree::Transform::Identity().asHomogeneousTransform().data());
+
+    bool ok = setRobotState(&fakeBase, &jointsPosSig, nullptr, nullptr, kinDyn.get());
```
Comment (S-Dafarra) [inline toolbox/src/RelativeTransform.cpp]: You can also use [``setJointPos``](http://wiki.icub.org/codyco/dox/html/idyntree/html/classiDynTree_1_1KinDynComputations.html#a55c2232d8d0069af6a3e80d0f901be02).

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,6 +5,11 @@
 cmake_minimum_required(VERSION 3.5)
 project(WB-Toolbox LANGUAGES CXX VERSION 3.1)
 
+if(WBTOOLBOX_BUILD_DOCS)
+    add_subdirectory(doc)
+    return()
+endif()
```
Comment (traversaro) [inline CMakeLists.txt]: I like this!

File: toolbox/src/RelativeTransform.cpp (hunk 1 above)
Comment (diegoferigo) [reply toolbox/src/RelativeTransform.cpp]: Despite the name is the same as `KinDynComputations::setRobotState`, it is not :sweat_smile: This method is [`WBBlock::setRobotState`](https://robotology.github.io/wb-toolbox/devel/doxygen/classwbt_1_1_w_b_block.html#a990354bac76a31d3baba9cf92c4b96cd).

Comment (S-Dafarra) [reply toolbox/src/RelativeTransform.cpp]: Ah dammit I did not notice it. Eventually you could wrap also that one :grin:

**Comment by traversaro** [comment]
Meta-comment, not directly related to this PR even if it is related to my documentation comments to the PR. 

Regarding documentation, it is not clear to me if:
* **option1**: we want to clearly let know to the users the YARP/iDynTree function/methods that a given block is wrapping, or
* **option2**: we want to consider WB-Toolbox an abstraction layer decoupled from YARP/iDynTree for have the freedom in the future to port it to other motion control interfaces or robot dynamics libraries. 

For  **option1** writing the documentation is easier, as the user will simply know that if he want more details he can check on the related YARP/iDynTree docs, and the WB-Toolbox docs just need to discuss how this function are wrapper, for example documenting the conversion from and to radians and degrees. For  **option2** documentation is much more complex, as in theory it should cover every detailed on the implementation.  

**My personal opinion:**
Given our experience with the `wholebodyinterface` and the `yarpwholebodyinterface` (as discussed in the JOSER paper) and the fact that the blocks are already designed to match YARP's motion interface semantics my preference is for **option1**. If in the future we will need to control robots with a different motion control interface, I think the easier options will be either to write a bridge Motion Control YARP Device, or implement new separate Simulink blocks that match the semantics of the different motion control interface.

**Comment by traversaro** [comment]
Just curiosity: why WB-Toolbox 4 instead of 3.2 ?

**Comment by diegoferigo** [comment]
@traversaro I think your question is directly related to #96. After the direction we took with the release of `3.0` (dropping `wholeBodyInterface` and migrate everything to `iDynTree::KinDynKinematics` and `yarp::dev::RemoteControlBoardRemapper`), my personal opinion would be to go towards what you described as **option 1**. I would like to split the core part of the toolbox that allows to easily wrap algorithms in Simulink and generate C++ code exploiting Simulink Coder. Then, developers that want to create a new toolbox can use this new project as base, and `wb-toolbox` would become an example of this new process for wrapping `Yarp` and `iDynTree`. We still have to plan a f2f meeting on this.

> Just curiosity: why WB-Toolbox 4 instead of 3.2 ?

The current `devel` branch contains a first draft of the new [`CONTRIBUTING`](https://github.com/robotology/wb-toolbox/blob/devel/.github/CONTRIBUTING.md#versioning) guidelines, where you can find also versioning information. Feedback is always accepted.

File: toolbox/src/RelativeTransform.cpp (hunk 1 above)
Comment (diegoferigo) [reply toolbox/src/RelativeTransform.cpp]: Just for curiosity, do you have to configure somehow the `world_H_base` transform in order to use `setJointPos`? From the documentation is written that it uses the previous value, but what if it has never been set?

cc @traversaro

Comment (traversaro) [reply toolbox/src/RelativeTransform.cpp]: I think the initial value is always identity, but I am not sure this is documented.

File: toolbox/matlab/library/WBToolboxLibrary_repository.mdl [hunk 3]
```diff
@@ -5029,6 +5155,165 @@ Library {
...
+	      ZOrder		      67
+	      RequestExecContextInheritance off
+	      Variant		      off
+	      Object {
+		$PropName		"MaskObject"
+		$ObjectID		225
+		$ClassName		"Simulink.Mask"
+		Type			"Relative Transform"
+		Description		"This block calculates the relative transform between two frames.\n\nAssuming DoFs is the number of deg"
+		"rees of freedom of the robot:\n\nInput:\n- Joint configuration: Vector of size DoFs, representing the configuration "
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: Updated all blocks specifying _internal_.

File: toolbox/matlab/library/WBToolboxLibrary_repository.mdl [hunk 4]
```diff
@@ -5029,6 +5155,165 @@ Library {
...
+		"of the joints.\n\nOutput:\n- Transform: a 4x4 matrix representing the homogenous transformation between frame1 and f"
+		"rame2.\n\nParameters:\n- Frame1 name: the name of the first frame. It should be part of the URDF model.\n- Frame2 na"
+		"me: the name of the second frame. It should be part of the URDF model."
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: What do you mean? The output port of the block is labeled as follows:
spectacle u31001

[the remark this answers is not in the document]
Comment (traversaro) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: Ah cool, I had no idea on this from the raw `.mdl` file.

File: toolbox/matlab/library/WBToolboxLibrary_repository.mdl [hunk 5]
```diff
@@ -3118,19 +3117,153 @@ Library {
...
+	  ZOrder		  44
+	  RequestExecContextInheritance	off
+	  Variant		  off
+	  Object {
+	    $PropName		    "MaskObject"
+	    $ObjectID		    133
+	    $ClassName		    "Simulink.Mask"
+	    Type		    "Set Motor References"
+	    Description		    "This block sets motor references. The type of control mode is specified as a parameter in the b"
+	    "lock.\n\nAssuming N as the number of motors (usually equal to DoFs, the number of joints):\n\nInput:\n- Reference"
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: You think is more correct documenting that _usually_ number of motors = number of DoFs?

File: toolbox/matlab/library/WBToolboxLibrary_repository.mdl [hunk 6]
```diff
@@ -5603,6 +5888,145 @@ Library {
...
+	  BackgroundColor	  "[0.513700, 0.851000, 0.670600]"
+	  RequestExecContextInheritance	off
+	  Variant		  off
+	  Object {
+	    $PropName		    "MaskObject"
+	    $ObjectID		    266
+	    $ClassName		    "Simulink.Mask"
+	    Type		    "Get Measurement"
+	    Description		    "This block gathers the measurement of the specified type.\n\nAssuming DoFs is the number of deg"
+	    "rees of freedom of the robot,\n\nOutput:\n* References: Vector of size DoFs, representing the requested measureme"
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: I have no idea to be honest. [The Yarp interface](http://yarp.it/classyarp_1_1dev_1_1ICurrentControl.html#a00aef7aa5c0d0da111cdf19ab3bad630) is not documented either.

[the remark this answers is not in the document]
Comment (traversaro) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: In http://yarp.it/classyarp_1_1dev_1_1ICurrentControl.html#a00aef7aa5c0d0da111cdf19ab3bad630 I read `pointer to the array that will store the output. Values are expressed in amperes.`

[the remark this answers is not in the document]
Comment (diegoferigo) [reply toolbox/matlab/library/WBToolboxLibrary_repository.mdl]: Woops, morning blindness :sweat_smile:

**Pull request merged**